libp2p/transports/wstransport.nim

```diff
@@ -0,0 +1,189 @@
+       ../stream/connection,
+       ../upgrademngrs/upgrade,
+       ws/ws
+
+logScope:
+  topics = "libp2p wstransport"
+
+export transport
+
+const
+  WsTransportTrackerName* = "libp2p.wstransport"
+
+type
+  WsStream = ref object of Connection
+    session: WSSession
+
+proc init*(T: type WsStream,
+           session: WSSession,
+           dir: Direction,
+           timeout = 10.minutes,
+           observedAddr: MultiAddress = MultiAddress()): T =
+
+  let stream = T(session: session,
+             timeout: timeout,
+             dir: dir,
+             observedAddr: observedAddr)
+  stream.initStream()
+  return stream
+
+method readOnce*(
+  s: WsStream,
+  pbytes: pointer,
+  nbytes: int):
+  Future[int] =
+  return s.session.recv(pbytes, nbytes) 
+
+method write*(s: WsStream, msg: seq[byte]):
+  Future[void] {.async.} =
+  #TODO the write procs can raise LPStreamEOF
+  # BUT this is not indicated in .raises. bc of async
+  # So this function can't just return the Future of ws.send
+  # We should add {.raises.} to every write, but this will probably
+  # break many things
+  try:
+    await s.session.send(msg, Opcode.Binary)
+  except WSClosedError:
+    raise newLPStreamEOFError()
+
+type
+  WsTransport* = ref object of Transport
+    httpserver: HttpServer
+    wsserver: WSServer
+
+    tlsPrivateKey: TLSPrivateKey
+    tlsCertificate: TLSCertificate
+
+
+method start*(
+  self: WsTransport,
+  ma: MultiAddress) {.async.} =
+  ## listen on the transport
+  ##
+
+  if self.running:
+    trace "WS transport already running"
+    return
+
+  await procCall Transport(self).start(ma)
+  trace "Starting WS transport"
+
+  self.httpserver = if isNil(self.tlsPrivateKey):
+     HttpServer.create(self.ma.initTAddress().tryGet())
+  else:
+    TlsHttpServer.create(self.ma.initTAddress().tryGet(),
+    self.tlsPrivateKey,
+    self.tlsCertificate)
+
+  self.wsserver = WSServer.new()
+
+  # always get the resolved address in case we're bound to 0.0.0.0:0
+  self.ma = MultiAddress.init(self.httpserver.sock.getLocalAddress()).tryGet()
+  self.running = true
+
+  trace "Listening on", address = self.ma
+
+method stop*(self: WsTransport) {.async, gcsafe.} =
+  ## stop the transport
+  ##
+
+  self.running = false # mark stopped as soon as possible
+
+  try:
+    trace "Stopping WS transport"
+    await procCall Transport(self).stop() # call base
+
+    # server can be nil
+    if not isNil(self.httpserver):
+      self.httpserver.stop()
+      await self.httpserver.closeWait()
+
+    self.httpserver = nil
+    trace "Transport stopped"
+  except CatchableError as exc:
+    trace "Error shutting down ws transport", exc = exc.msg
+
+method accept*(self: WsTransport): Future[Connection] {.async, gcsafe.} =
+  ## accept a new WS connection
+  ##
+
+  if not self.running:
+    raise newTransportClosedError()
+
+  try:
+    let transp = await self.httpserver.accept()
+    #TODO could this be blocking? If so, someone could stuck our accept loop
```

Comment:
chronos is completely async and you'd need to go out of your way (e.g. use `waitFor()`) to combine it with sync code, so should be easy to spot. The only caveat is long CPU intensive tasks that don't yield to back to the main event loop.

Reply:
I meant blocking on the sense of this accept loop,
Eg, if someone start to send the http request, but doesn't follow up on a ws connection or send a malicious request, could we get stuck between accept and handle? Thus, blocking other peers from connecting

Reply:
Put in an other way, for the duration of `handleRequest`, we don't accept other peers since there is a single accept loop. So, if a malicious/bad behaving peer can make `handleRequest` slow, he could potentially block us from accepting other peers.
Thus, either `handleRequest` should always finish in a timely manner, or we should implement a better thing here to allow multiple connection requests in parallel.

This was more a note to myself, to check if `handleRequest` is instantaneous or if I need to allow parallel connection attempts :)

Reply:
Ah, that's a good question and there are several layers of indirection here. 

First there is the TCP backlog (100 in chronos by default), which means that the OS (or TCP stack) will queue up to backlog size incoming connections, this connections will time out eventually, but they won't immediately drop giving it a chance of being picked up at some point. 

The second layer is that we can have several concurrent accepts running at the same time, we don't currently do this in the switch because we haven't seen an issue with incoming connections, but this is certainly possible.

In any case, I wouldn't actually worry about this in the transport itself, the calling application should be able to decide what's a good number of concurrent accepts to run, whether it should periodically timeout, etc... This move the complexity up to a layer that can actually make informed decisions about it.

Reply:
Yes, but the specificity of WS is that the accept is a two phase process in the application, whereas for TCP for instance, when `accept` returns, you know you have a valid & ready connection
So while for TCP, having a single accept loop is sufficient because a bad behaving peer can't block it, for WS we need to consider the possibility of having the accept loop blocked.

Anyway, the only `await`s in `handleRequest` are `send`s, so I don't think a bad behaving peer could block `handleRequest`, since in the worst case, the OS will buffer the send, so no issue here AFAICT :)

Reply:
Yes, by the time we've got to `handleRequest` we've already read all the data (since it's an HTTP request and headers aren't streamed), but we can still have a malicious attacker simply not finishing writing the headers, so I think a timeout when reading headers would be appropriate - this however should go in `nim-ws`.

Reply:
Moved upstream: https://github.com/status-im/nim-websock/issues/79